### ISSUE TITLE
✍️ Contribution period: <Batya> 

### ISSUE THREAD

**Issue by kbatya** [ersilia-os/ersilia#836]
### Week 1 - Get to know the community

- [X] Join the communication channels
- [X] Open a GitHub issue (this one!)
- [x] Install the Ersilia Model Hub and test the simplest model
- [x] Write a motivation statement to work at Ersilia
- [x] Submit your first contribution to the Outreachy site

### Week 2 - Install and run an ML model

- [x] Select a model from the suggested list
- [x] Install the model in your system
- [x] Run predictions for the EML
- [x] Compare results with the Ersilia Model Hub implementation!
- [x] Install and run Docker!

### Week 3 - Propose new models

- [x] Suggest a new model and document it (1)
- [x] Suggest a new model and document it (2)
- [x] Suggest a new model and document it (3)

### Week 4 - Prepare your final application

- [x] Submit the final application in the Outreachy website

**Comment by DhanshreeA** [comment]
Hi @kbatya can you confirm if you were able to run Ersilia with the simplest model and the output is as mentioned in the instructions?

**Comment by kbatya** [comment]
I confirm that I installed Ersilia ran the simplest model and got the output as written in the instructions.  I should to describe my steps, I want to do it now.

**Comment by kbatya** [comment]
1. I joined the Slack workspace and introduced myself in the general channel
2.  I opened the Git Issue #836

**Comment by kbatya** [comment]
3. Installation: I work on windows, and I've installed Ubuntu terminal. Because I forgot the Administrator password, I couldn't run sudo command. So I must reset the password: 
I open Windows Command Promt and wrote commands: 
wsl -d Ubuntu -u root # launch Ubuntu in recovery mode and in root access
passwd <WSL username> # prompted with a new password that I entered and confirmed
exit

**Comment by kbatya** [comment]
3. Installation: Pre-requisites: 
     1. gcc compiler
     2. Python and Miniconda
     3. Git and GitHub CLI
     4. Git LFS
     5. isaura data lake:
                     I tried the code "conda activate Ercilia" and got error. So I think in the instructions it should be added creation of environment and not a notation that is explained below.  So I created the environment "Ersilia" and activated it, installed isaura 
     6. Docker
    Installation Ersilia"
      1. I didn't create and didn't activated  the environment because I did it in previouse step when installed isaura data lake
          but install Ersilia Python package
       2. I checked that it works 
                ersilia --version
               output: ersilia, version 0.1.27

**Comment by kbatya** [comment]
Model usage:
# fetch model from remote repository using slug ...
ersilia fetch retrosynthetic-accessibility

# serve model
ersilia serve retrosynthetic-accessibility

# make prediction
ersilia run -i "C1=C(SC(=N1)SC2=NN=C(S2)N)[N+](=O)[O-]"
ouput: 
  File "/home/lplane/ersilia/ersilia/io/readers/file.py", line 321, in read_input_columns
    if len(h) == 1:
TypeError: object of type 'NoneType' has no len()

I reinstalled ersilia without isaura dat lake installation
now the output was: 
{
    "input": {
        "key": "NQQBNZBOOHHVQP-UHFFFAOYSA-N",
        "input": "C1=C(SC(=N1)SC2=NN=C(S2)N)[N+](=O)[O-]",
        "text": "C1=C(SC(=N1)SC2=NN=C(S2)N)[N+](=O)[O-]"
    },
    "output": {
        "outcome": [
            0.9924924
        ]
    }
}

**Comment by kbatya** [comment]
I am an experienced software engineer with a passion for data science. I discovered the exciting world of data science five years ago and completed an intensive course in machine learning and deep learning in 2020. I have implemented models using Python and Conda to gain a profound understanding of the field. However, I have been struggling to find a job in this field due to my lack of experience and belonging to a discriminated group. I am part of the five percent of people with similar characteristics who work as data scientists. Since the beginning of my career, I have been looking for meaningful work that has a positive impact on humanity, saving lives, as medicine does. At the same time, I am passionate about technology and innovation. The Ersilia project aligns with all my interests and goals: it provides an opportunity to make a significant contribution, sharpen my data science skills, and be an active participant in an open-source environment.

**Comment by kbatya** [comment]
Week 2:
I've choose "Plasma Protein Binding (IDL-PPBopt)" because I like both the application - prediction of Human Plasma Protein Binding and the algorithm - Deep Learning method

**Comment by DhanshreeA** [comment]
Thank you for the updates @kbatya

**Comment by kbatya** [comment]
Install the model in my system:
1. clone from github to my computer
`
git clone https://github.com/Louchaofeng/IDL-PPBopt
cd IDL-PPBopt
conda create --name IDL-PPBopt python=3.7
conda activate `IDL-PPBopt`

**Comment by kbatya** [comment]
[Requirements](https://github.com/Louchaofeng/IDL-PPBopt/blob/main/README.md#requirements)
python 3.7
pytorch 1.5.0
openbabel 2.4.1
rdkit
scikit learn
scipy
cairosvg

**Comment by kbatya** [comment]
**Run** IDL-PPBopt.ipynb in jupyter notebook(first install it):
```
pip install jupyter notebook
cd Code
jupyter notebook IDL-PPBopt.ipynb
```

**Comment by kbatya** [comment]
**Task 4: understand Ersilia's backend:**
I found the model Human Plasma Protein Binding (PPB) of Compounds in the Ersilia Model Hub:
Identifiers

- EOS model ID: eos22io

- Slug: idl-ppbopt

**Comment by DhanshreeA** [comment]
> Run IDL-PPBopt.ipynb in jupyter notebook: When I ran cell "Step 4. Predict values" I got an error [error_log.txt](https://github.com/ersilia-os/ersilia/files/12871459/error_log.txt) I suggest it is because of Incompatibility between versions of Cuda, my computer has Cuda 11.7 and I need Cuda 10.1 for the model. I installed pytorch with cuda 10.1. Still, I get the error. @DhanshreeA @mentors

Hey @kbatya if you have Cuda 11.7 but the model needs Cuda 10.1, wouldn't it make sense to instead work with a different Cuda driver ie 10.1 in addition to installing pytorch with cuda 10.1?

**Comment by kbatya** [comment]
> > Run IDL-PPBopt.ipynb in jupyter notebook: When I ran cell "Step 4. Predict values" I got an error [error_log.txt](https://github.com/ersilia-os/ersilia/files/12871459/error_log.txt) I suggest it is because of Incompatibility between versions of Cuda, my computer has Cuda 11.7 and I need Cuda 10.1 for the model. I installed pytorch with cuda 10.1. Still, I get the error. @DhanshreeA @mentors
> 
> Hey @kbatya if you have Cuda 11.7 but the model needs Cuda 10.1, wouldn't it make sense to instead work with a different Cuda driver ie 10.1 in addition to installing pytorch with cuda 10.1?

Thanks, I will try

**Comment by HellenNamulinda** [comment]
Hello @kbatya, 
Just following up, were you able to try running the model with the recommended CUDA version?
Is there any way we can support you?

**Comment by kbatya** [comment]
> Hello @kbatya, Just following up, were you able to try running the model with the recommended CUDA version? Is there any way we can support you?

Hello @HellenNamulinda ,
thank you for your concern
I've tried to install Cuda driver 10.1,10.2 and compotable torch with cuda 10.1,10.2 but I had the same error 
[error_log.txt](https://github.com/ersilia-os/ersilia/files/12916066/error_log.txt)
Now I am trying to run without cuda - CPU version., meantime I have other errors

**Comment by kbatya** [comment]
- I've solved the above bug . In the Jupyter Notebook in the second cell "Related Functions" I didn't replace "torch.cuda.LongTensor" with "torch.LongTensor"

- I copied eml_canonical.csv file with a dataset for prediction 

`sudo cp /home/lplane/ersilia/notebooks/eml_canonical.csv ./Code/`

**Comment by DhanshreeA** [comment]
Hi @kbatya thank you for the updates. Good job debugging so far, and getting the model to run. Let us know how it goes when you run the model against EML and then compare it with Ersilia's implementation against EML. Afterwards, you can move onto week 3 tasks.

**Comment by kbatya** [comment]
### Compare results with the Ersilia Model Hub implementation!
jupyter notebook **IDL-PPBopt_EML.ipynb** After the dataset file "eml_canonical.csv" was loaded into a dataframe I renamed the column name 'can_smiles' into 'cano_smiles' to match the 
structure of the model's input:
`smiles_tasks_df.rename(columns={'can_smiles':'cano_smiles'},inplace=True)`

after running the model in step 4.Predicted values
I uploaded the csv file with prediction values from running the model from Ersilia Hub: output_eml.csv
then I loaded the file into dataframe :
`ersilia_predictions_df = pd.read_csv('output_eml.csv')`
After I explored the **ersilia model** predictions I saw:
number of input smiles is 442
number of predictions is 433
After I explored the **original model** predictions I saw:
number of input smiles is 442
number of predictions is 433

That means there are 9 input smiles that wasn't predicted by both models:
after i run the code to select them 
`
I got:
<table border="1" class="dataframe">
  <thead>
    <tr style="text-align: right;">
      <th></th>
      <th>key</th>
      <th>input</th>
      <th>PPB_VALUES</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <th>60</th>
      <td>FAQLAUHZSGTTLN-UHFFFAOYSA-N</td>
      <td>[CaH2]</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>170</th>
      <td>KRHYYFGTRYWZRS-UHFFFAOYSA-M</td>
      <td>[F-]</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>213</th>
      <td>ZCYVEMRRCGMTRW-UHFFFAOYSA-N</td>
      <td>[I]</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>307</th>
      <td>XLYOFNOQVPJJNP-UHFFFAOYSA-N</td>
      <td>O</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>332</th>
      <td>WCUXLLCKKVVCTQ-UHFFFAOYSA-M</td>
      <td>[Cl-].[K+]</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>333</th>
      <td>NLKNQRATVPKPDG-UHFFFAOYSA-M</td>
      <td>[K+].[I-]</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>338</th>
      <td>RWSOTUBLDIXVET-UHFFFAOYSA-N</td>
      <td>S</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>374</th>
      <td>FJKGRAZQBBWYLG-UHFFFAOYSA-M</td>
      <td>N.N.[F-].[Ag+]</td>
      <td>NaN</td>
    </tr>
    <tr>
      <th>378</th>
      <td>FAPWRFPIFSIZLT-UHFFFAOYSA-M</td>
      <td>[Na+].[Cl-]</td>
      <td>NaN</td>
    </tr>
  </tbody>
</table>

**Comment by kbatya** [comment]
This model predicts the binding affinity of the plasma protein with the compound.
<table border="1">
<caption><h1>Affinity of the binding of the plasma protein</h1></caption>
  <thead>
    <tr>
<th>Level of the affinity </th>
      <th>PPB values(Predictions)</th>
     
    </tr>
  </thead>
  <tbody>
    <tr>
       <td>High</td>
      <td>0.8-1</td>
    </tr>
    <tr>
      <td>Medium</td>
      <td>0.4-0.79</td>
          </tr>
    <tr>
      <td>Low</td>
      <td>0.01-0.39</td>
      </tr>
    <tr>
      <td>Not featured</td>
      <td>Null</td>
       </tr>
  </tbody>
</table>

**Comment by kbatya** [comment]
********### Week 3 - Propose new models
**Suggest a new model and document it (1)**
- **Model name**
Molecular Transformer - Reaction Prediction

- **Publication**
[https://arxiv.org/pdf/1811.02633v2.pdf](url)

- **Source code**
[https://github.com/pschwllr/MolecularTransformer](url)

          **Requirements:**
            1. six
            2. tqdm
            3. torch>=0.4.1
            4. torchtext=0.3.1
            5. future
- **Mode**
Pretrained
- **Description**
The model takes SMILES strings of reagents and reactants and with a multi-head attention Molecular Transformer predicts the chemical reaction's products.
- **Input**
Compounds
- **Output**
Compounds

- **Relevancy for the Ersilia project**

1. **Acceleration of the drug discovery process**. Instead of relying solely on labor-intensive experimental trials, this approach allows researchers to quickly explore a wide range of reactions and compound combinations.
2. **Cost reduction of the drug discovery process**. The model can help prioritize reactions and identify potential candidates for further study, saving valuable resources. 
3. **Targeting Neglected Diseases**: For orphan and neglected diseases, there may be limited funding and resources available for research. Leveraging AI to predict reactions and screen compounds can help focus efforts on the most promising candidates.
In summary, using a Molecular Transformer with multi-head attention for predicting chemical reaction products aligns with Ersilia's mission by providing a powerful tool for accelerating drug discovery, making it more cost-effective, and prioritizing compounds with potential for neglected diseases, will be helpful in low- and middle-income countries (LMIC).

**Comment by kbatya** [comment]
### Week 3 - Propose new models
**Suggest a new model and document it (2)**
- **Model name**
SMILES-X

- **Publication**
[SMILES-X: autonomous molecular compounds characterization for small datasets without descriptors](https://iopscience.iop.org/article/10.1088/2632-2153/ab57f3/pdf)

- **Source code - GitHub**
[SMILES-X](https://github.com/Lambard-ML-Team/SMILES-X)

**Requirements:**
 * SMILES-X works on CPU, but it is highly recommended to have an access to GPU:

1. CUDA == 10.1
2. cuDNN == 8.0.3
3. RDKit
4. python == 3.7
5. pandas == 1.3.5
6. numpy == 1.18.5
7. matplotlib == 3.5.2
8. GPy == 1.10.0
9. GPyOpt == 1.2.6
10. scikit-learn == 1.0.2
11. adjustText == 0.8
12. scipy == 1.7.3
13. tensorflow == 2.1.3

- **Mode**
    Pretrained
- **Description**
The SMILES-X is an autonomous pipeline that finds the best neural architectures to predict a physicochemical property from molecular SMILES only. No human-engineered descriptors are needed. The SMILES-X has been specially designed for small datasets (<< 1000 samples).
- **Input**
Compounds
- **Output**
Compounds

- **Relevancy for the Ersilia project**
Autonomous Neural Network's Architecture Search will speed up the discovery of drugs, while saving time and resources.
- **License**
MIT

**Comment by kbatya** [comment]
### Week 3 - Propose new models
**Suggest a new model and document it (3)**
- **Model name**
ReLeaSE

- **Publication**
[Deep reinforcement learning for de novo drug design](https://www.science.org/doi/10.1126/sciadv.aap7885#sec-4)

- **Source code - GitHub**
[ReLeaSE](https://github.com/isayev/ReLeaSE)

**Requirements:**

1. Modern NVIDIA GPU, compute capability 3.5 of newer.
2. CUDA 9.0
3. Pytorch 0.4.1
4. Tensorflow 1.8.0 with GPU support
5. RDKit
6. Scikit-learn
7. Numpy
8. tqdm
9. Mordred

- **Mode**
    Pretrained
- **Description**
ReLeaSE is a novel computational strategy for designing molecules using deep and reinforcement learning. It combines generative and predictive neural networks to create targeted chemical libraries based on simplified molecular-input line-entry system (SMILES) strings. The method involves separate training of generative and predictive models with supervised learning, followed by joint training with reinforcement learning to bias molecule generation towards desired properties. In a proof-of-concept study, ReLeaSE was used to design chemical libraries optimized for specific properties, such as structural complexity, melting point, or inhibitory activity against Janus protein kinase 2. This approach has broad applications for generating tailored chemical libraries with single or multiple desired properties.
- **Input**
Compounds
- **Output**
Compounds

- **Relevancy for the Ersilia project**
Autonomous Neural Network's Architecture Search will speed up the discovery of drugs, while saving time and resources.
- **License**
MIT

**Comment by GemmaTuron** [comment]
Hello,

Thanks for your work during the Outreachy contribution period, we hope you enjoyed it! We will now close this issue while we work on the selection of interns. Thanks again!
